Update to Python 3.6+

smrg-lm opened the pull request:
Other changes (none in the math code):

- Rename of 'python' folder to package name 'pmlib'.
- Package structure files.
- Python style: same indentation in all files and pep8 (mostly).
- Code readability (most of the refactoring).
- Prefer simple names when they are not equivocal (e.g boundaryCond -> boundary).
- Simplification of boundary names (improveds code and usability).
- Added NOTE, BUG and FIXME comments.
- Removed systemSetup.py (it can be run as ./python -m pmlib -s 22050 net.json mod.json).

smrg-lm committed:
Update to Python 3.6+

Other changes (none in the math code):

- Rename of 'python' folder to package name 'pmlib'.
- Package structure files.
- Python style: same indentation in all files and pep8 (mostly).
- Code readability (most of the refactoring).
- Prefer simple names when they are not equivocal (e.g boundaryCond -> boundary).
- Simplification of boundary names (improveds code and usability).
- Added NOTE, BUG and FIXME comments.
- Removed systemSetup.py (it can be run as ./python -m pmlib -s 22050 net.json mod.json).

smrg-lm commented:
Hi

I was trying your library and needed to update it to python 3, then I saw some things that could be improved while I was watching the code. Well, the thing is I ended up doing maintenance work because I like pm synthesis and this is missing in sc, and you seem to be a mathematician, and the examples sound very nice. If you want to move forward this library I can help you with some code design (I can't do the math or maintain or improve them). 
Also, I'm currently working in this project https://github.com/smrg-lm/sc3 because the power of python make possible libraries like this one with scipy and alike. That's all.

There is one more change not specifically listed. I changed the long names of clamped, simply or free ends, I believe is much more easy to type a simple code than the names, even easy to remember.

I hope you like the update.

michaeldzjap commented:
@smrg-lm Thanks for your PR. I haven't actively developed this project in quite some years. Looking at it back now there would have been a few things I would have approached differently, one of which definitely would be coding style and adding some tests for instance.

With regards to moving forward with this library; I don't think I can really commit to this myself unfortunately. I would be a bit hesitant to accept any major overhauls in functionality, but I am open to any PR's that improve the existing code base. You can alway decide to fork this project and branch of from there if you feel the need to make any more involved adjustments of course.

With regards to your proposed updates and merging your PR; I looked at the changes briefly and I think I can find myself in most (if not all) of them, but I'd like to take a closer look before merging. This might take a while, since I am quite busy with work and other things at the moment, but I will try to make some time for this somewhere in the coming weeks.

smrg-lm commented:
I understand, just wanted to share it back to you, please don't feel committed.

Changes are just in format not in content, the examples from the documentation run exactly as before. I had to edit so many lines to add spaces and indentation to make it more pep8 friendly. I needed to change only a few lines to make it compatible with Python 3, checked the rest, and formated some messages with f strings because looks more clear to me.

The code of systemSetup is __main__ now, imports are relative, and __init__ is just for convention, so it can be used as a package or invoked as program from the command line. The structure is the same it was before.

michaeldzjap commented on pmlib/ResonatorNetwork.py:
```diff
@@ -0,0 +1,412 @@
+
+from math import pi
+from numbers import Number
+import warnings
```
```suggestion
```

Imported, but not used

michaeldzjap commented on pmlib/ResonatorNetwork.py:
```diff
@@ -0,0 +1,412 @@
+from scipy.linalg import solve
+from scipy.linalg import eig
+
+import scipy as sp
```
```suggestion
```

Imported, but not used

michaeldzjap commented on pmlib/Resonator2D.py:
```diff
@@ -0,0 +1,77 @@
+
+from math import sqrt
+
+import numpy as np
```
```suggestion
```

Imported, but unused.

michaeldzjap commented on pmlib/Resonator2D.py:
```diff
@@ -0,0 +1,77 @@
+from math import sqrt
+
+import numpy as np
+from scipy.sparse import dia_matrix,identity
```
```suggestion
from scipy.sparse import identity
```

scipy.sparse.dia_matrix imported, but unused.

michaeldzjap commented on pmlib/Resonator2D.py:
```diff
@@ -0,0 +1,77 @@
+                 boundary='SSSS', epsilon=1):
+        super().__init__(gamma, kappa, b1, b2, boundary)
+        self._epsilon = epsilon
+        # self.h = None
```
I think these commented out lines can be deleted?

smrg-lm replied:
Yes, they were there just to mark that those are attributes created later.

michaeldzjap commented on pmlib/ResonatorBase.py:
```diff
@@ -0,0 +1,73 @@
+
```
```suggestion
```

michaeldzjap commented on ResonatorNetwork.sc:
```diff
@@ -4,7 +4,7 @@ ResonatorNetwork {
 	*initClass {
 		Class.initClassTree(String);
-		pythonScriptPath = this.class.filenameSymbol.asString.dirname ++ "/python";
+		pythonScriptPath = this.class.filenameSymbol.asString.dirname;
```
The _modalData.json_ and _networkArgs.json_ files are now placed in the root of the project and git will interpret them as untracked files due to this change, so they should probably be added to .gitignore.

The whole "old" _python_ sub directory could be removed from .gitignore actually.

smrg-lm replied:
About modalData and networkArgs, I thought it could be better to use Platform.defaultTempDir if the files aren't needed or the path is not set to an specific directory, but I didn't make changes in sclang classes if it was no needed for python.

smrg-lm replied:
I believe this suggested change was added, I can't see ``++ "/python"`` in my local clone.

michaeldzjap reviewed (changes requested):
Couple of minor suggestions. I haven't looked at all changes in detail. I've tested that everything still works functionality wise, and it seems that it does. That is most important to me for now.

Ideally, I think a lot of the code could use a more thorough clean up and refactoring if I am honest. Not sure what I was thinking when I now look back at how certain things are coded and structured... Hopefully I can make time for this at some point.

Anyway, thanks for you PR. If you can fix the outstanding things (all should be pretty trivial) then I am happy to merge this.

smrg-lm committed:
Update pmlib/ResonatorNetwork.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update pmlib/Resonator2D.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update pmlib/Resonator2D.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update pmlib/ResonatorBase.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update pmlib/ResonatorNetwork.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update .gitignore.

smrg-lm committed:
Removed commented dynamic attributes.

smrg-lm committed:
A more correct solution to the hardcoded sampling rate issue.

Fixes #1

smrg-lm commented:
I added a patch to fix issue #1 in a more consistent way. I removed default values for ``calc_modes`` since they aren't used and to avoid setting them to variable data.

smrg-lm commented:
Force push because I did silly mistakes with the commit.

michaeldzjap commented on ResonatorNetwork.sc:
```diff
@@ -92,11 +92,12 @@ ResonatorNetwork {
 	}
 
-	calcModalData { arg minFreq=25,maxFreq=(Server.default.sampleRate ? 44100).div(2),minT60=0.01,gain=1,pathname=pythonScriptPath ++ "/modalData.json",incl="ynnn",async=false;
+	calcModalData { arg minFreq=25,maxFreq=22050,minT60=0.01,gain=1,pathname=pythonScriptPath ++ "/modalData.json",incl="ynnn",async=false;
```
Could you elaborate on why you decided to change `(Server.default.sampleRate ? 44100).div(2)` to just 22050? Granted, the fact that `Server.default` doesn't necessarily correspond to the server that is used to run the simulation and the fact that the default server has to be booted doesn't make it and ideal solution, but I still think it is better than just assuming that the sample rate is fixed to 44100.

smrg-lm replied:
Mostly because what you say at the end and because I'm not sure if it's good idea to use expressions for args default values. It can be reverted, will work fine.

michaeldzjap commented on pmlib/Resonator1D.py:
```diff
@@ -0,0 +1,73 @@
+
```
```suggestion
```

smrg-lm replied:
Ok. This was a convention that I use when there is no comment at the top, maybe should avoid it from now on.

michaeldzjap commented on pmlib/Resonator2D.py:
```diff
@@ -0,0 +1,72 @@
+
```
```suggestion
```

michaeldzjap commented on pmlib/__main__.py:
```diff
@@ -0,0 +1,55 @@
+
```
```suggestion
```

michaeldzjap commented on pmlib/sparse_add.py:
```diff
@@ -0,0 +1,358 @@
+
```
```suggestion
```

michaeldzjap commented on pmlib/sparse_diff_matr_1d.py:
```diff
@@ -1,65 +1,71 @@
+
```
```suggestion
```

michaeldzjap commented on pmlib/sparse_diff_matr_2d.py:
```diff
@@ -0,0 +1,109 @@
+
```
```suggestion
```

michaeldzjap commented on .gitignore:
```diff
@@ -2,9 +2,5 @@
 *networkTest.scd
 # ignore all pyc files
 *.pyc
-# ignore everything inside tests directory
```
I think _networkArgs.json_ and _modalData.json_ should be added here. Otherwise they will appear as untracked files to _Git_.

smrg-lm replied:
I think it could be better to use Platform.defaultTempDir if the files aren't needed later or the path is not set to an specific directory, or use home as default. Because the user is writing data to the code directory and also is a bit hidden there.

michaeldzjap replied:
Right, that would definitely work for _networkArgs.json_. The problem is that _modalData.json_ I do want to keep sometimes. Especially for large networks it can take a while to do the computation (minutes). So then it is kind of nice to check if the file already exists and if so, not having to run the offline computation part again. But writing data to the code directory isn't an ideal solution for sure. Even more because of the fact that _modalData.json_ is not uniquely identifiable for instance. If I generate it for a particular network and want to cache it, but then do the computation for another network it will overwrite _modalData.json_ again. So the whole way _modalData.json_ is handled should probably be rethought to be honest, but that is going to be more work of course.

So in order to not break any existing / expected behaviour I would prefer to just add these 2 files to _.gitignore_, even though I agree this could be handled better.

smrg-lm replied:
Very sorry for the late reply, I forgot about it. If you want you can resolve this two issues it's ok by me and can't commit more changes.

smrg-lm committed:
Update pmlib/Resonator1D.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update pmlib/Resonator2D.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update pmlib/__main__.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update pmlib/sparse_add.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update pmlib/sparse_diff_matr_1d.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>

smrg-lm committed:
Update pmlib/sparse_diff_matr_2d.py

Co-authored-by: Michael Dzjaparidze <michael.dzjaparidze@gmail.com>
